src/Http/Routing/src/Builder/EndpointRoutingApplicationBuilderExtensions.cs

```diff
@@ -99,7 +99,10 @@ public static IApplicationBuilder UseEndpoints(this IApplicationBuilder builder,
             var routeOptions = builder.ApplicationServices.GetRequiredService<IOptions<RouteOptions>>();
             foreach (var dataSource in endpointRouteBuilder.DataSources)
             {
-                routeOptions.Value.EndpointDataSources.Add(dataSource);
+                if (!routeOptions.Value.EndpointDataSources.Contains(dataSource))
```

Comment:
I was going to say this makes the algorithm O(N^2) but this is the number of datasources, not endpoints so that's likely to be small. Still it's a small performance regression.

I wonder if there's a way to mark the endpoint route builder itself as already processed instead.

cc @javiercn